src/platform/backends/shared/base_vm_image_vault.h

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2020-2021 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef MULTIPASS_BASE_VM_IMAGE_VAULT_H
+#define MULTIPASS_BASE_VM_IMAGE_VAULT_H
+
+#include <multipass/format.h>
+#include <multipass/query.h>
+#include <multipass/vm_image.h>
+#include <multipass/vm_image_host.h>
+#include <multipass/vm_image_vault.h>
+
+#include <string>
+#include <vector>
+
+namespace multipass
+{
+class BaseVMImageVault : public VMImageVault
+{
+public:
+    explicit BaseVMImageVault(const std::vector<VMImageHost*>& image_hosts)
+        : image_hosts{image_hosts}, remote_image_host_map{vault::configure_image_host_map(image_hosts)} {};
+
+    VMImageHost* image_host_for(const std::string& remote_name) const override
+    {
+        auto it = remote_image_host_map.find(remote_name);
+        if (it == remote_image_host_map.end())
+        {
+            throw std::runtime_error(
+                fmt::format("Remote \'{}\' is not found. Please use `multipass find` for supported remotes and images.",
+                            remote_name));
+        }
+
+        return it->second;
+    };
+
+    std::vector<VMImageInfo> all_info_for(const Query& query) const override
+    {
+        std::vector<VMImageInfo> images_info;
+
+        if (!query.remote_name.empty())
+        {
+            auto image_host = image_host_for(query.remote_name);
+            images_info = image_host->all_info_for(query);
+        }
+        else
+        {
+            for (const auto& image_host : image_hosts)
+            {
+                auto images_info = image_host->all_info_for(query);
+
+                if (!images_info.empty())
+                {
+                    break;
+                }
+            }
+        }
+
+        if (images_info.empty())
+            throw std::runtime_error(fmt::format("Unable to find an image matching \"{}\"", query.release));
```

Comment:
I had a cursory look but wasn't able to find where this was before. Do we really need to throw in this case?

Reply:
Nevermind, I found it.